proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,873 @@
+
+An `xdstp://` URI specifies a logic authority and it’s necessary to map to some physical
+network address during the resource fetch process. In the existing xDS configuration
+fetch configuration, this capability is provided by a `ConfigSource` delivered alongside a
+resource name in the xDS resource graph.
+
+This proposal adds to `ConfigSource` the ability to specify the
+authorities for which the `ConfigSource` may be used:
+
+```
+message Authority {
+  string name = 1;
+
+  … when signing is added, items such as CA trust chain, cert pinning …
+}
+
+message ConfigSource {
+  repeated Authority authorities = N;
+
+  …  extant ConfigSource …
+}
+```
+
+When the `ConfigSource` adjacent to a resource name matches the URI authority,
+the `ConfigSource` will continue to be used.
+
+When there is no match, the xDS client will fallback to the bootstrap to map
+from authority to `ConfigSource`. The bootstrap format is client specific, an
+example of this is:
+
+```
+// List of authorities and config.
+repeated ConfigSource xdstp_config_sources = N;
+
+```
+
+A client may support only static bootstrap configuration of `ConfigSource` for authority mapping. It is
+expected that relevant authorities are configured in the bootstrap for referencable servers.
+
+To support ADS with multiple control planes, the
+[`ApiType`](https://github.com/envoyproxy/envoy/blob/31128e7dc22355876020188bc8feb99304663041/api/envoy/config/core/v3/config_source.proto#L44)
+enum in `ApiConfigSource` will be augmented with `AGGREGATED_DELTA_GRPC`.
+This will replace and deprecate the existing `ads` field in `ConfigSource`. When
+used in `xdstp://` authority resolution, an `ConfigSource` will act as an ADS
+transport if configured with this new enum value. Instead of pointing to a
+single shared ADS `ConfigSource` declared in the bootstrap, any resource
+authority (regardless of type) that maps to a `ConfigSource` with
+`AGGREGATED_DELTA_GRPC` will be multiplexed on the xDS stream specified by the
+`ConfigSource`.
+
+### Discovery request and responses
+
+Clients are expected to have knowledge ahead of time (via mechanisms not part of
+this proposal, e.g. support documentation) on whether the servers specified in
+the bootstrap supports the new URI conventions.
+
+Examples below are provided for delta xDS, but this proposal does not limit
+usage to delta xDS. We directionally intend for delta xDS to surplant SotW xDS
+over time, the use of inlined entries in list collections allows simple
+management servers that do not wish to perform state tracking to perform
+state-of-the-world updates. This is achieved by sending the entire state, e.g.
+every listener inlined in a `ListenerCollection`. However, migration is
+expected to be slow on the uptake and continued SotW support is provided for
+this proposal.
+
+### Server redirects
+
+The server may issue redirects by including the following message
+in a `Resource` wrapped in the `Any` message:
+
+```
+message Redirect {
```

Comment:
One more quick change here.  Let's change this to:

```
message ResourceLocator {
  string name = 1;
}
```

That way, we can use the same message for both redirects and for subscriptions/unsubscriptions, since we'll be making a similar change for subscriptions/unsubscriptions as part of the dynamic parameter design.